File: csharp/1457.pseudo-palindromic-paths-in-a-binary-tree.cs
/*
 * @lc app=leetcode id=1457 lang=csharp
 *
 * [1457] Pseudo-Palindromic Paths in a Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {

    private static int PseudoPalindromicPaths(TreeNode root, int state, int count) {
        if (root == null) {
            return 0;
        }
        int m = 1 << root.val;
        state ^= m;
        if ((state & m) == 0) {
            --count;
        } else {
            ++count;
        }
        if (root.left == null && root.right == null) {
            return count <= 1 ? 1 : 0;
        }
        return PseudoPalindromicPaths(root.left, state, count) + PseudoPalindromicPaths(root.right, state, count);
    }

    public int PseudoPalindromicPaths (TreeNode root) {
        return PseudoPalindromicPaths(root, 0, 0);
    }
}
// @lc code=end

File: java/1457.pseudo-palindromic-paths-in-a-binary-tree.java
/*
 * @lc app=leetcode id=1457 lang=java
 *
 * [1457] Pseudo-Palindromic Paths in a Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pseudoPalindromicPaths (TreeNode root) {
        return pseudoPalindromicPaths(root, 0);
    }

    private static int pseudoPalindromicPaths(TreeNode curr, int acc) {
        if (curr == null) {
            return 0;
        }
        acc ^= (1 << curr.val);
        if (curr.left == null && curr.right == null) {
            int c = 0;
            while (acc != 0) {
                c += acc & 1;
                acc >>= 1;
            }
            return c <= 1 ? 1 : 0;
        }
        return pseudoPalindromicPaths(curr.left, acc) + pseudoPalindromicPaths(curr.right, acc);
    }
}
// @lc code=end

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
